# Error Analysis: Prediction Structure

- **Question:** Does future-related syntactic structure reliably indicate a prediction?

In [ ]:
import os
import sys

import pandas as pd

notebook_dir = os.getcwd()

sys.path.append(os.path.join(notebook_dir, '../'))

from data_processing import DataProcessing
from feature_extraction import SpacyFeatureExtraction

In [2]:
pd.set_option('max_colwidth', 800)
# pd.set_option('display.max_columns', None)
# pd.set_option('display.max_columns', 40)
# pd.set_option('display.max_rows', None)

## Load Data

In [3]:
base_data_path = DataProcessing.load_base_data_path(notebook_dir)
# combine_data_path = os.path.join(base_data_path, 'financial_phrase_bank/combined_generated_fin_phrase_bank')
train_data = os.path.join(base_data_path, 'classification_results/train_synthetic-v1_2026-03-23/seed3')

In [4]:
# model_results_path = os.path.join(combine_data_path, 'inference_chronicle2050_2026-03-07_21-51-47.csv')
test_data = os.path.join(train_data, 'external_fpb-maya-binary-imbalanced-96d-v1/', 'ml_classifiers_fpb-maya-binary-imbalanced-96d-v1.csv')

model_results_df = DataProcessing.load_from_file(test_data, 'csv', sep=',')
cols_to_drop = ["Dataset Name", "Author Type", "maya_label"]
compare_y_vs_yhats_df = DataProcessing.drop_df_columns(model_results_df, cols_to_drop)
compare_y_vs_yhats_df

,Base Sentence,Sentence Label,Base Sentence Embedding,perceptron,sgd_classifier,logistic_regression,ridge_classifier,decision_tree_classifier,random_forest_classifier,gradient_boosting_classifier,support_vector_machine_classifier,x_gradient_boosting_classifier
0,With the new production plant the company would increase its capacity to meet the expected increase in demand and would improve the use of raw materials and therefore increase the production profitability .,1,[-1.22568220e-01 2.48624131e-01 -4.82655168e-02 -1.09789923e-01\n -8.58348906e-02 3.63291465e-02 -1.27964811e-02 7.17335939e-02\n -1.18364163e-01 2.27930522e+00 -3.32476139e-01 -1.46164745e-02\n 1.25499114e-01 7.25146085e-02 6.57393187e-02 -5.76901138e-02\n -2.05907249e-03 1.51917112e+00 -3.29480022e-01 -4.57408801e-02\n 2.22120117e-02 1.05339624e-01 -4.22692373e-02 -1.23566784e-01\n -5.79119613e-03 6.18153140e-02 -5.48806489e-02 2.96797305e-02\n -1.07008472e-01 -9.98812243e-02 -9.39770564e-02 3.87728140e-02\n -5.82127571e-02 7.14930072e-02 9.65250358e-02 -3.75642404e-02\n 7.18023349e-03 6.21852428e-02 7.68979266e-02 -1.69069305e-01\n 1.44475931e-03 7.71993026e-02 3.41709815e-02 -1.10992551e-01\n -1.81387588e-02 -8.61631632e-02 -2.34903637e-02 6.00658916e-02\n -3....,1,1,1,1,0,1,1,1,1
1,"According to the company 's updated strategy for the years 2009-2012 , Basware targets a long-term net sales growth in the range of 20 % -40 % with an operating profit margin of 10 % -20 % of net sales .",1,[-2.63320029e-01 3.28540474e-01 4.48558182e-02 -1.09422831e-02\n 3.39385755e-02 -1.25387803e-01 2.09340192e-02 8.32975954e-02\n 1.10356383e-01 1.74510765e+00 -3.98153275e-01 1.41264707e-01\n -8.98099765e-02 -1.47727029e-02 5.77163734e-02 -5.09223454e-02\n -4.27257568e-02 1.41277170e+00 -2.09747717e-01 1.56779997e-02\n -3.64826852e-03 5.97854443e-02 -7.89400712e-02 -1.39893517e-02\n 1.13595366e-01 9.83026922e-02 -5.14178127e-02 7.06899390e-02\n 2.47105230e-02 1.06609218e-01 -4.19648038e-03 -1.85196903e-02\n -1.48957130e-02 9.90139246e-02 1.10973224e-01 -2.59710420e-02\n -6.08825274e-02 4.48305123e-02 4.36691009e-02 -9.27021950e-02\n 6.36729002e-02 1.98753644e-02 2.10891932e-01 -8.54000822e-02\n 4.67465706e-02 -2.04764325e-02 -3.69948000e-02 1.16207667e-01\n 1....,1,1,0,0,1,1,1,0,1
2,TeliaSonera TLSN said the offer is in line with its strategy to increase its ownership in core business holdings and would strengthen Eesti Telekom 's offering to its customers .,1,[ 1.15080299e-02 1.98225155e-01 1.95062440e-02 -5.97744621e-02\n 1.75585955e-01 -1.25474662e-01 -3.72287892e-02 -2.83286646e-02\n -3.56662273e-02 1.96045303e+00 -2.79843062e-01 -1.04898110e-01\n 4.49508205e-02 1.31481076e-02 6.55795187e-02 -4.11130209e-03\n -5.56710083e-03 1.11790729e+00 -1.21459611e-01 -9.75316912e-02\n 2.18506586e-02 1.38311490e-01 9.17155594e-02 -5.85136563e-02\n 1.03342809e-01 7.32071251e-02 -3.92427929e-02 1.13294438e-01\n -1.07247204e-01 5.58346435e-02 -9.28281173e-02 5.48683619e-03\n -1.84886660e-02 5.24056628e-02 4.05881368e-02 -8.35432187e-02\n 1.13391867e-02 2.78038122e-02 -5.44880610e-03 -1.02035753e-01\n -6.69988245e-02 7.79658780e-02 5.37769347e-02 -1.39616817e-01\n 2.96867881e-02 1.13102477e-02 -3.53476144e-02 -1.88835058e-02\n -4....,1,1,1,1,1,1,1,1,1
3,"STORA ENSO , NORSKE SKOG , M-REAL , UPM-KYMMENE Credit Suisse First Boston ( CFSB ) raised the fair value for shares in four of the largest Nordic forestry groups .",1,[-5.94305294e-03 1.57299116e-01 -1.01663664e-01 -4.74716797e-02\n 1.79840446e-01 2.63532978e-02 4.07044962e-02 -3.17418873e-02\n -1.32127828e-03 1.28309309e+00 -3.01563025e-01 4.19499092e-02\n 4.76229452e-02 -1.16947450e-01 -1.42535036e-02 2.37643998e-02\n -3.57760116e-02 8.38545859e-01 -5.43218441e-02 2.42432952e-03\n 7.67285563e-03 4.88467440e-02 4.74129207e-02 -2.67546187e-04\n 6.47759363e-02 7.66171236e-03 -7.71588413e-03 6.14660196e-02\n 9.25298501e-03 1.29891783e-01 6.76145032e-02 -7.35822739e-03\n 3.22291143e-02 1.50113806e-01 1.32441163e-01 -9.81159508e-03\n -1.3941115

In [5]:
model_col_names = compare_y_vs_yhats_df.columns.to_list()[3:]
model_col_names, len(model_col_names)

(['perceptron',
  'sgd_classifier',
  'logistic_regression',
  'ridge_classifier',
  'decision_tree_classifier',
  'random_forest_classifier',
  'gradient_boosting_classifier',
  'support_vector_machine_classifier',
  'x_gradient_boosting_classifier'],
 9)

In [6]:
compare_y_vs_mv_df = compare_y_vs_yhats_df[model_col_names].mode(axis=1)
compare_y_vs_yhats_df['Majority Vote'] = compare_y_vs_mv_df
compare_y_vs_yhats_df

,Base Sentence,Sentence Label,Base Sentence Embedding,perceptron,sgd_classifier,logistic_regression,ridge_classifier,decision_tree_classifier,random_forest_classifier,gradient_boosting_classifier,support_vector_machine_classifier,x_gradient_boosting_classifier,Majority Vote
0,With the new production plant the company would increase its capacity to meet the expected increase in demand and would improve the use of raw materials and therefore increase the production profitability .,1,[-1.22568220e-01 2.48624131e-01 -4.82655168e-02 -1.09789923e-01\n -8.58348906e-02 3.63291465e-02 -1.27964811e-02 7.17335939e-02\n -1.18364163e-01 2.27930522e+00 -3.32476139e-01 -1.46164745e-02\n 1.25499114e-01 7.25146085e-02 6.57393187e-02 -5.76901138e-02\n -2.05907249e-03 1.51917112e+00 -3.29480022e-01 -4.57408801e-02\n 2.22120117e-02 1.05339624e-01 -4.22692373e-02 -1.23566784e-01\n -5.79119613e-03 6.18153140e-02 -5.48806489e-02 2.96797305e-02\n -1.07008472e-01 -9.98812243e-02 -9.39770564e-02 3.87728140e-02\n -5.82127571e-02 7.14930072e-02 9.65250358e-02 -3.75642404e-02\n 7.18023349e-03 6.21852428e-02 7.68979266e-02 -1.69069305e-01\n 1.44475931e-03 7.71993026e-02 3.41709815e-02 -1.10992551e-01\n -1.81387588e-02 -8.61631632e-02 -2.34903637e-02 6.00658916e-02\n -3....,1,1,1,1,0,1,1,1,1,1
1,"According to the company 's updated strategy for the years 2009-2012 , Basware targets a long-term net sales growth in the range of 20 % -40 % with an operating profit margin of 10 % -20 % of net sales .",1,[-2.63320029e-01 3.28540474e-01 4.48558182e-02 -1.09422831e-02\n 3.39385755e-02 -1.25387803e-01 2.09340192e-02 8.32975954e-02\n 1.10356383e-01 1.74510765e+00 -3.98153275e-01 1.41264707e-01\n -8.98099765e-02 -1.47727029e-02 5.77163734e-02 -5.09223454e-02\n -4.27257568e-02 1.41277170e+00 -2.09747717e-01 1.56779997e-02\n -3.64826852e-03 5.97854443e-02 -7.89400712e-02 -1.39893517e-02\n 1.13595366e-01 9.83026922e-02 -5.14178127e-02 7.06899390e-02\n 2.47105230e-02 1.06609218e-01 -4.19648038e-03 -1.85196903e-02\n -1.48957130e-02 9.90139246e-02 1.10973224e-01 -2.59710420e-02\n -6.08825274e-02 4.48305123e-02 4.36691009e-02 -9.27021950e-02\n 6.36729002e-02 1.98753644e-02 2.10891932e-01 -8.54000822e-02\n 4.67465706e-02 -2.04764325e-02 -3.69948000e-02 1.16207667e-01\n 1....,1,1,0,0,1,1,1,0,1,1
2,TeliaSonera TLSN said the offer is in line with its strategy to increase its ownership in core business holdings and would strengthen Eesti Telekom 's offering to its customers .,1,[ 1.15080299e-02 1.98225155e-01 1.95062440e-02 -5.97744621e-02\n 1.75585955e-01 -1.25474662e-01 -3.72287892e-02 -2.83286646e-02\n -3.56662273e-02 1.96045303e+00 -2.79843062e-01 -1.04898110e-01\n 4.49508205e-02 1.31481076e-02 6.55795187e-02 -4.11130209e-03\n -5.56710083e-03 1.11790729e+00 -1.21459611e-01 -9.75316912e-02\n 2.18506586e-02 1.38311490e-01 9.17155594e-02 -5.85136563e-02\n 1.03342809e-01 7.32071251e-02 -3.92427929e-02 1.13294438e-01\n -1.07247204e-01 5.58346435e-02 -9.28281173e-02 5.48683619e-03\n -1.84886660e-02 5.24056628e-02 4.05881368e-02 -8.35432187e-02\n 1.13391867e-02 2.78038122e-02 -5.44880610e-03 -1.02035753e-01\n -6.69988245e-02 7.79658780e-02 5.37769347e-02 -1.39616817e-01\n 2.96867881e-02 1.13102477e-02 -3.53476144e-02 -1.88835058e-02\n -4....,1,1,1,1,1,1,1,1,1,1
3,"STORA ENSO , NORSKE SKOG , M-REAL , UPM-KYMMENE Credit Suisse First Boston ( CFSB ) raised the fair value for shares in four of the largest Nordic forestry groups .",1,[-5.94305294e-03 1.57299116e-01 -1.01663664e-01 -4.74716797e-02\n 1.79840446e-01 2.63532978e-02 4.07044962e-02 -3.17418873e-02\n -1.32127828e-03 1.28309309e+00 -3.01563025e-01 4.19499092e-02\n 4.76229452e-02 -1.16947450e-01 -1.42535036e-02 2.37643998e-02\n -3.57760116e-02 8.38545859e-01 -5.43218441e-02 2.42432952e-03\n 7.67285563e-03 4.88467440e-02 4.74129207e-02 -2.67546187e-04\n 6.47759363e-02 7.66171236e-03 -7.71588413e-03 6.14660196e-02\n 9.25298501e-03 1.29891783e-01 6.76145032e-02 -7.35822739e-03\n 3.22291143e-02 1.50113806e-01 1.32441163e-01 -9.81159

In [7]:
model_col_names

['perceptron',
 'sgd_classifier',
 'logistic_regression',
 'ridge_classifier',
 'decision_tree_classifier',
 'random_forest_classifier',
 'gradient_boosting_classifier',
 'support_vector_machine_classifier',
 'x_gradient_boosting_classifier']

In [8]:
meta_data = compare_y_vs_yhats_df.iloc[:, 2:-1]
meta_data

,Base Sentence Embedding,perceptron,sgd_classifier,logistic_regression,ridge_classifier,decision_tree_classifier,random_forest_classifier,gradient_boosting_classifier,support_vector_machine_classifier,x_gradient_boosting_classifier
0,[-1.22568220e-01 2.48624131e-01 -4.82655168e-02 -1.09789923e-01\n -8.58348906e-02 3.63291465e-02 -1.27964811e-02 7.17335939e-02\n -1.18364163e-01 2.27930522e+00 -3.32476139e-01 -1.46164745e-02\n 1.25499114e-01 7.25146085e-02 6.57393187e-02 -5.76901138e-02\n -2.05907249e-03 1.51917112e+00 -3.29480022e-01 -4.57408801e-02\n 2.22120117e-02 1.05339624e-01 -4.22692373e-02 -1.23566784e-01\n -5.79119613e-03 6.18153140e-02 -5.48806489e-02 2.96797305e-02\n -1.07008472e-01 -9.98812243e-02 -9.39770564e-02 3.87728140e-02\n -5.82127571e-02 7.14930072e-02 9.65250358e-02 -3.75642404e-02\n 7.18023349e-03 6.21852428e-02 7.68979266e-02 -1.69069305e-01\n 1.44475931e-03 7.71993026e-02 3.41709815e-02 -1.10992551e-01\n -1.81387588e-02 -8.61631632e-02 -2.34903637e-02 6.00658916e-02\n -3....,1,1,1,1,0,1,1,1,1
1,[-2.63320029e-01 3.28540474e-01 4.48558182e-02 -1.09422831e-02\n 3.39385755e-02 -1.25387803e-01 2.09340192e-02 8.32975954e-02\n 1.10356383e-01 1.74510765e+00 -3.98153275e-01 1.41264707e-01\n -8.98099765e-02 -1.47727029e-02 5.77163734e-02 -5.09223454e-02\n -4.27257568e-02 1.41277170e+00 -2.09747717e-01 1.56779997e-02\n -3.64826852e-03 5.97854443e-02 -7.89400712e-02 -1.39893517e-02\n 1.13595366e-01 9.83026922e-02 -5.14178127e-02 7.06899390e-02\n 2.47105230e-02 1.06609218e-01 -4.19648038e-03 -1.85196903e-02\n -1.48957130e-02 9.90139246e-02 1.10973224e-01 -2.59710420e-02\n -6.08825274e-02 4.48305123e-02 4.36691009e-02 -9.27021950e-02\n 6.36729002e-02 1.98753644e-02 2.10891932e-01 -8.54000822e-02\n 4.67465706e-02 -2.04764325e-02 -3.69948000e-02 1.16207667e-01\n 1....,1,1,0,0,1,1,1,0,1
2,[ 1.15080299e-02 1.98225155e-01 1.95062440e-02 -5.97744621e-02\n 1.75585955e-01 -1.25474662e-01 -3.72287892e-02 -2.83286646e-02\n -3.56662273e-02 1.96045303e+00 -2.79843062e-01 -1.04898110e-01\n 4.49508205e-02 1.31481076e-02 6.55795187e-02 -4.11130209e-03\n -5.56710083e-03 1.11790729e+00 -1.21459611e-01 -9.75316912e-02\n 2.18506586e-02 1.38311490e-01 9.17155594e-02 -5.85136563e-02\n 1.03342809e-01 7.32071251e-02 -3.92427929e-02 1.13294438e-01\n -1.07247204e-01 5.58346435e-02 -9.28281173e-02 5.48683619e-03\n -1.84886660e-02 5.24056628e-02 4.05881368e-02 -8.35432187e-02\n 1.13391867e-02 2.78038122e-02 -5.44880610e-03 -1.02035753e-01\n -6.69988245e-02 7.79658780e-02 5.37769347e-02 -1.39616817e-01\n 2.96867881e-02 1.13102477e-02 -3.53476144e-02 -1.88835058e-02\n -4....,1,1,1,1,1,1,1,1,1
3,[-5.94305294e-03 1.57299116e-01 -1.01663664e-01 -4.74716797e-02\n 1.79840446e-01 2.63532978e-02 4.07044962e-02 -3.17418873e-02\n -1.32127828e-03 1.28309309e+00 -3.01563025e-01 4.19499092e-02\n 4.76229452e-02 -1.16947450e-01 -1.42535036e-02 2.37643998e-02\n -3.57760116e-02 8.38545859e-01 -5.43218441e-02 2.42432952e-03\n 7.67285563e-03 4.88467440e-02 4.74129207e-02 -2.67546187e-04\n 6.47759363e-02 7.66171236e-03 -7.71588413e-03 6.14660196e-02\n 9.25298501e-03 1.29891783e-01 6.76145032e-02 -7.35822739e-03\n 3.22291143e-02 1.50113806e-01 1.32441163e-01 -9.81159508e-03\n -1.39411157e-02 1.04931206e-03 -1.90248825e-02 -9.27351974e-03\n -2.62032263e-02 7.86117762e-02 8.33134279e-02 2.64521260e-02\n -1.58701912e-02 9.79095325e-03 3.53994928e-02 -1.72072705e-02\n 7....,0,0,0,0,1,0,0,0,0
4,[-9.75671411e-02 2.17742935e-01 -5.24893478e-02 2.29859371e-02\n 1.95262451e-02 -3.96593288e-02 -2.93270722e-02 1.45730842e-02\n 8.01267326e-02 2.14558053e+00 -3.02207738e-01 7.73812085e-02\n 9.75400433e-02 -5.86019419e-02 2.85570417e-02 -6.01695329e-02\n -6.48141513e-03 1.10976875e+00 -1.62209153e-01 -2.63800696e-02\n 6.90186918e-02 5.40496930e-02 -4.89046909e-02 -1.06065851e-02\n 3.81638743e-02 8.63624960e-02 -1.46540195e-01 -7.51399994e-02\n -5.42312972e-02 -2.43689902e-02 3.15490589e-02 -7.77387097e-02\n -1.04405805e-01 1.55526847e-01 2.55947262e-02 -2.01004483e-02\n -4.3354682

In [9]:
meta_data = compare_y_vs_yhats_df.iloc[:, 3:-1]
compare_y_vs_yhats_df['Meta Data'] = meta_data.to_dict(orient='records')
compare_y_vs_yhats_df
y_vs_yhats_df = DataProcessing.drop_df_columns(compare_y_vs_yhats_df, meta_data.columns.to_list())
y_vs_yhats_df


,Base Sentence,Sentence Label,Base Sentence Embedding,Majority Vote,Meta Data
0,With the new production plant the company would increase its capacity to meet the expected increase in demand and would improve the use of raw materials and therefore increase the production profitability .,1,[-1.22568220e-01 2.48624131e-01 -4.82655168e-02 -1.09789923e-01\n -8.58348906e-02 3.63291465e-02 -1.27964811e-02 7.17335939e-02\n -1.18364163e-01 2.27930522e+00 -3.32476139e-01 -1.46164745e-02\n 1.25499114e-01 7.25146085e-02 6.57393187e-02 -5.76901138e-02\n -2.05907249e-03 1.51917112e+00 -3.29480022e-01 -4.57408801e-02\n 2.22120117e-02 1.05339624e-01 -4.22692373e-02 -1.23566784e-01\n -5.79119613e-03 6.18153140e-02 -5.48806489e-02 2.96797305e-02\n -1.07008472e-01 -9.98812243e-02 -9.39770564e-02 3.87728140e-02\n -5.82127571e-02 7.14930072e-02 9.65250358e-02 -3.75642404e-02\n 7.18023349e-03 6.21852428e-02 7.68979266e-02 -1.69069305e-01\n 1.44475931e-03 7.71993026e-02 3.41709815e-02 -1.10992551e-01\n -1.81387588e-02 -8.61631632e-02 -2.34903637e-02 6.00658916e-02\n -3....,1,"{'perceptron': 1, 'sgd_classifier': 1, 'logistic_regression': 1, 'ridge_classifier': 1, 'decision_tree_classifier': 0, 'random_forest_classifier': 1, 'gradient_boosting_classifier': 1, 'support_vector_machine_classifier': 1, 'x_gradient_boosting_classifier': 1}"
1,"According to the company 's updated strategy for the years 2009-2012 , Basware targets a long-term net sales growth in the range of 20 % -40 % with an operating profit margin of 10 % -20 % of net sales .",1,[-2.63320029e-01 3.28540474e-01 4.48558182e-02 -1.09422831e-02\n 3.39385755e-02 -1.25387803e-01 2.09340192e-02 8.32975954e-02\n 1.10356383e-01 1.74510765e+00 -3.98153275e-01 1.41264707e-01\n -8.98099765e-02 -1.47727029e-02 5.77163734e-02 -5.09223454e-02\n -4.27257568e-02 1.41277170e+00 -2.09747717e-01 1.56779997e-02\n -3.64826852e-03 5.97854443e-02 -7.89400712e-02 -1.39893517e-02\n 1.13595366e-01 9.83026922e-02 -5.14178127e-02 7.06899390e-02\n 2.47105230e-02 1.06609218e-01 -4.19648038e-03 -1.85196903e-02\n -1.48957130e-02 9.90139246e-02 1.10973224e-01 -2.59710420e-02\n -6.08825274e-02 4.48305123e-02 4.36691009e-02 -9.27021950e-02\n 6.36729002e-02 1.98753644e-02 2.10891932e-01 -8.54000822e-02\n 4.67465706e-02 -2.04764325e-02 -3.69948000e-02 1.16207667e-01\n 1....,1,"{'perceptron': 1, 'sgd_classifier': 1, 'logistic_regression': 0, 'ridge_classifier': 0, 'decision_tree_classifier': 1, 'random_forest_classifier': 1, 'gradient_boosting_classifier': 1, 'support_vector_machine_classifier': 0, 'x_gradient_boosting_classifier': 1}"
2,TeliaSonera TLSN said the offer is in line with its strategy to increase its ownership in core business holdings and would strengthen Eesti Telekom 's offering to its customers .,1,[ 1.15080299e-02 1.98225155e-01 1.95062440e-02 -5.97744621e-02\n 1.75585955e-01 -1.25474662e-01 -3.72287892e-02 -2.83286646e-02\n -3.56662273e-02 1.96045303e+00 -2.79843062e-01 -1.04898110e-01\n 4.49508205e-02 1.31481076e-02 6.55795187e-02 -4.11130209e-03\n -5.56710083e-03 1.11790729e+00 -1.21459611e-01 -9.75316912e-02\n 2.18506586e-02 1.38311490e-01 9.17155594e-02 -5.85136563e-02\n 1.03342809e-01 7.32071251e-02 -3.92427929e-02 1.13294438e-01\n -1.07247204e-01 5.58346435e-02 -9.28281173e-02 5.48683619e-03\n -1.84886660e-02 5.24056628e-02 4.05881368e-02 -8.35432187e-02\n 1.13391867e-02 2.78038122e-02 -5.44880610e-03 -1.02035753e-01\n -6.69988245e-02 7.79658780e-02 5.37769347e-02 -1.39616817e-01\n 2.96867881e-02 1.13102477e-02 -3.53476144e-02 -1.88835058e-02\n -4....,1,"{'perceptron': 1, 'sgd_classifier': 1, 'logistic_regression': 1, 'ridge_classifier': 1, 'decision_tree_classifier': 1, 'random_forest_classifier': 1, 'gradient_boosting_classifier': 1, 'support_vector_machine_classifier': 1, 'x_gradient_boosting_classifier': 1}"
3,"STORA ENSO , NORSKE SKOG , M-REAL , UPM-KYMMENE Credit Suisse First Boston ( CFSB ) raised the fair value for shares in four of the largest Nordic forestry groups .",1,[-5.94305294e-03 1.572

## Get class of each sentence

- [ ] True Label vs Majority Vote
- [ ] True Label vs Each Model

## True Label vs Majority Vote

In [10]:
def classification_of_mv_sentence(df):
    copy_df = df.copy()
    copy_df['Class'] = None

    for idx, row in copy_df.iterrows():
        y = row['Sentence Label']
        y_hat = row['Majority Vote']

        # y = 1 (positive class)
        if y == 1:
            if y_hat == 1:  # TP
                copy_df.loc[idx, 'Class'] = "TP"
            else:  # FN
                copy_df.loc[idx, 'Class'] = "FN"

        # y = 0 (negative class)
        elif y == 0:
            if y_hat == 0:  # TN
                copy_df.loc[idx, 'Class'] = "TN"
            else:  # FP
                copy_df.loc[idx, 'Class'] = "FP"

    return copy_df

In [11]:
mv_with_class_df = classification_of_mv_sentence(y_vs_yhats_df)
mv_with_class_df

,Base Sentence,Sentence Label,Base Sentence Embedding,Majority Vote,Meta Data,Class
0,With the new production plant the company would increase its capacity to meet the expected increase in demand and would improve the use of raw materials and therefore increase the production profitability .,1,[-1.22568220e-01 2.48624131e-01 -4.82655168e-02 -1.09789923e-01\n -8.58348906e-02 3.63291465e-02 -1.27964811e-02 7.17335939e-02\n -1.18364163e-01 2.27930522e+00 -3.32476139e-01 -1.46164745e-02\n 1.25499114e-01 7.25146085e-02 6.57393187e-02 -5.76901138e-02\n -2.05907249e-03 1.51917112e+00 -3.29480022e-01 -4.57408801e-02\n 2.22120117e-02 1.05339624e-01 -4.22692373e-02 -1.23566784e-01\n -5.79119613e-03 6.18153140e-02 -5.48806489e-02 2.96797305e-02\n -1.07008472e-01 -9.98812243e-02 -9.39770564e-02 3.87728140e-02\n -5.82127571e-02 7.14930072e-02 9.65250358e-02 -3.75642404e-02\n 7.18023349e-03 6.21852428e-02 7.68979266e-02 -1.69069305e-01\n 1.44475931e-03 7.71993026e-02 3.41709815e-02 -1.10992551e-01\n -1.81387588e-02 -8.61631632e-02 -2.34903637e-02 6.00658916e-02\n -3....,1,"{'perceptron': 1, 'sgd_classifier': 1, 'logistic_regression': 1, 'ridge_classifier': 1, 'decision_tree_classifier': 0, 'random_forest_classifier': 1, 'gradient_boosting_classifier': 1, 'support_vector_machine_classifier': 1, 'x_gradient_boosting_classifier': 1}",TP
1,"According to the company 's updated strategy for the years 2009-2012 , Basware targets a long-term net sales growth in the range of 20 % -40 % with an operating profit margin of 10 % -20 % of net sales .",1,[-2.63320029e-01 3.28540474e-01 4.48558182e-02 -1.09422831e-02\n 3.39385755e-02 -1.25387803e-01 2.09340192e-02 8.32975954e-02\n 1.10356383e-01 1.74510765e+00 -3.98153275e-01 1.41264707e-01\n -8.98099765e-02 -1.47727029e-02 5.77163734e-02 -5.09223454e-02\n -4.27257568e-02 1.41277170e+00 -2.09747717e-01 1.56779997e-02\n -3.64826852e-03 5.97854443e-02 -7.89400712e-02 -1.39893517e-02\n 1.13595366e-01 9.83026922e-02 -5.14178127e-02 7.06899390e-02\n 2.47105230e-02 1.06609218e-01 -4.19648038e-03 -1.85196903e-02\n -1.48957130e-02 9.90139246e-02 1.10973224e-01 -2.59710420e-02\n -6.08825274e-02 4.48305123e-02 4.36691009e-02 -9.27021950e-02\n 6.36729002e-02 1.98753644e-02 2.10891932e-01 -8.54000822e-02\n 4.67465706e-02 -2.04764325e-02 -3.69948000e-02 1.16207667e-01\n 1....,1,"{'perceptron': 1, 'sgd_classifier': 1, 'logistic_regression': 0, 'ridge_classifier': 0, 'decision_tree_classifier': 1, 'random_forest_classifier': 1, 'gradient_boosting_classifier': 1, 'support_vector_machine_classifier': 0, 'x_gradient_boosting_classifier': 1}",TP
2,TeliaSonera TLSN said the offer is in line with its strategy to increase its ownership in core business holdings and would strengthen Eesti Telekom 's offering to its customers .,1,[ 1.15080299e-02 1.98225155e-01 1.95062440e-02 -5.97744621e-02\n 1.75585955e-01 -1.25474662e-01 -3.72287892e-02 -2.83286646e-02\n -3.56662273e-02 1.96045303e+00 -2.79843062e-01 -1.04898110e-01\n 4.49508205e-02 1.31481076e-02 6.55795187e-02 -4.11130209e-03\n -5.56710083e-03 1.11790729e+00 -1.21459611e-01 -9.75316912e-02\n 2.18506586e-02 1.38311490e-01 9.17155594e-02 -5.85136563e-02\n 1.03342809e-01 7.32071251e-02 -3.92427929e-02 1.13294438e-01\n -1.07247204e-01 5.58346435e-02 -9.28281173e-02 5.48683619e-03\n -1.84886660e-02 5.24056628e-02 4.05881368e-02 -8.35432187e-02\n 1.13391867e-02 2.78038122e-02 -5.44880610e-03 -1.02035753e-01\n -6.69988245e-02 7.79658780e-02 5.37769347e-02 -1.39616817e-01\n 2.96867881e-02 1.13102477e-02 -3.53476144e-02 -1.88835058e-02\n -4....,1,"{'perceptron': 1, 'sgd_classifier': 1, 'logistic_regression': 1, 'ridge_classifier': 1, 'decision_tree_classifier': 1, 'random_forest_classifier': 1, 'gradient_boosting_classifier': 1, 'support_vector_machine_classifier': 1, 'x_gradient_boosting_classifier': 1}",TP
3,"STORA ENSO , NORSKE SKOG , M-REAL , UPM-KYMMENE Credit Suisse First Boston ( CFSB ) raised the fair value for shares in four of the largest Nordic forestry groups .",1,[-5.943

| $ y $ | $ y_{hat} $ | Outcome |
|----------------|----------------|----------|
| 1              | 1              | **TP**   |
| 1              | 0              | **FN**  (miss out) |
| 0              | 0              | **TN**   |
| 0              | 1              | **FP**  (inflate) |

**Confusion Matrix (sklearn-style layout)**

| y \ y_hat |   0   |   1   |
|-----------|-------|-------|
| **0**     |  TN   |  FP  (inflate) |
| **1**     |  FN  (miss out) |  TP   |

## Linguistic Features

### Structures of a prediction

|  |   Investigation of Future Reference Expression in Trend Info by Nakajima et al., (2014)   |   spaCy   |
|-----------|-------|-------|
| 1     |  [Action]*[State Change]   |  AUX + VERB |

In [12]:
aux_verb_sentences = []
disable_components = []

class_names = mv_with_class_df['Class'].unique()

for class_name in class_names:
    print(f"\t\tCLASS: {class_name}")
    filt_class_name = (mv_with_class_df['Class'] == class_name)
    class_df = mv_with_class_df[filt_class_name]

    # EXTRACT FEATURES
    spacy_fe = SpacyFeatureExtraction(class_df, 'Base Sentence')
    # ner_df = spacy_fe.extract_ner_features(disable_components, visualize=False)
    pos_df = spacy_fe.extract_pos_features(disable_components, visualize=False)
    # print(pos_df['Term'].value_counts())
    # print(pos_df['Dependency'].value_counts())
    # print(pos_df['POS Label'].value_counts())

    # Future-related contains :: spaCy
    aux_sentences = set(pos_df[pos_df['POS Label'] == 'AUX']['Sentence'].tolist())
    verb_sentences = set(pos_df[pos_df['POS Label'] == 'VERB']['Sentence'].tolist())
    class_aux_verb_sentences = aux_sentences & verb_sentences

    for sentence in class_aux_verb_sentences:
        aux_verb_sentences.append({
            'Class': class_name,
            'Sentence': sentence
        })

    percentage_aux_verb_within_class = round(len(class_aux_verb_sentences) / len(class_df) * 100, 2)
    print(f"\n\t\tCLASS: {class_name}")
    print(f"Number of sentences with AUX + VERB: {len(class_aux_verb_sentences)} out of {len(class_df)} = {percentage_aux_verb_within_class}%")
    for sentence in list(class_aux_verb_sentences)[:3]:
        print(f"  - {sentence}")
    print("#" * 50)
    print()


		CLASS: TP

EXTRACTING POS FEATURES


Extracting POS Features: 100%|██████████| 388/388 [00:01<00:00, 305.16doc/s]



DONE EXTRACTING POS FEATURES

		CLASS: TP
Number of sentences with AUX + VERB: 319 out of 388 = 82.22%
  - Master of Mayawas jointly developed by Nokia Siemens Networks and UFA - FremantleMedia , and will be actively advertised by Maxis in the end of May 2007 .
  - The acquisition will considerably increase Kemira 's sales and market position in the Russian metal industry coatings market .
  - The share sale , to foreign and Finnish investors , is expected to raise some euro300 million ( US$ 380 million ) .
##################################################

		CLASS: FN

EXTRACTING POS FEATURES


Extracting POS Features: 100%|██████████| 70/70 [00:00<00:00, 264.28doc/s]



DONE EXTRACTING POS FEATURES

		CLASS: FN
Number of sentences with AUX + VERB: 50 out of 70 = 71.43%
  - The office space will rise above the remodeled Cannon Street underground station .
  - `` We are proud to welcome another distribution facility to north Mississippi and the region known as the logistics center of the United States , '' said Gray Swoope , executive director of MDA .
  - Mr. Atul Chopra , Chief Operating Officer & President , Tecnotree elaborated '' I am eager to see the continued success of the former Tecnomen , bringing Lifetree product to MENA , under the Tecnotree umbrella .
##################################################

		CLASS: FP

EXTRACTING POS FEATURES


Extracting POS Features: 100%|██████████| 1852/1852 [00:06<00:00, 295.04doc/s]



DONE EXTRACTING POS FEATURES

		CLASS: FP
Number of sentences with AUX + VERB: 1311 out of 1852 = 70.79%
  - Olvi has posted a strong set of figures for the first six months of this year .
  - Incap Contract Manufacturing will carry out the manufacturing for these agreements at its factory in Tumkur , near Bangalore .
  - The shares shall be repurchased through public trading , for which reason the shares are repurchased otherwise than in proportion to the holdings of the shareholders .
##################################################

		CLASS: TN

EXTRACTING POS FEATURES


Extracting POS Features: 100%|██████████| 2536/2536 [00:08<00:00, 283.43doc/s]



DONE EXTRACTING POS FEATURES

		CLASS: TN
Number of sentences with AUX + VERB: 1054 out of 2536 = 41.56%
  - The new factory working model and reorganisations would decrease Nokian Tyres ' costs in the factory by EUR 30 million ( USD 38.7 m ) .
  - The repurchase of the bonds was executed in the open market in accordance with section 7 f of the terms and conditions of the Convertible Bonds .
  - Sales in Latin America increased by 42 % to EUR 432mn , and a total of 8.7 mn mobile devices were sold in the area , representing an increase of 32 % from the corresponding period in 2009 .
##################################################



## Takeaways

1. **TP (correct predictions):** 82.22% contain AUX + VERB, so the future-related structure is a strong but NOT a full indicator of a prediction because ~18% are correctly identified without this structure.

2. **FN (missed predictions):** 71.43% contain AUX + VERB, yet the model still misclassified them as non-predictions. This suggests the models may have failed learning the explicit future-related structure in some cases.

3. **FP (inflated predictions):** 70.79% contain AUX + VERB — too close to FN. This shows that AUX + VERB alone $\nrightarrow$ prediction.

4. **TN (correct non-predictions):** 41.56% contain AUX + VERB — the lowest of all, so most non-predictions do NOT carry future-related structure.


In [13]:
# Create dataframe
aux_verb_df = pd.DataFrame(aux_verb_sentences)
aux_verb_df

,Class,Sentence
0,TP,"Master of Mayawas jointly developed by Nokia Siemens Networks and UFA - FremantleMedia , and will be actively advertised by Maxis in the end of May 2007 ."
1,TP,The acquisition will considerably increase Kemira 's sales and market position in the Russian metal industry coatings market .
2,TP,"The share sale , to foreign and Finnish investors , is expected to raise some euro300 million ( US$ 380 million ) ."
3,TP,"Later on , about 150 Finnish-made houses will be build in the area ."
4,TP,"In addition to fireplace exports , demand for lining stone has exceeded the level of the earlier part of the year and will continue to be clearly brisker for the remainder of the year ."
...,...,...
2729,TN,"POYRY PLC Additional information by : Martin Kuzaj , President , Industry Business Group , Finland Tel. +358 10 33 21179 Sanna Paivaniemi , Director , Investor Relations , Poyry PLC , Finland Tel. +358 10 33 23002 Poyry is a global consulting and engineering company dedicated to balanced sustainability ."
2730,TN,"The brokerage said 2006 has seen a ` true turning point ' in European steel base prices , with better pricing seen carrying through the second quarter of 2006 ."
2731,TN,"Asimilar agreement was signed between China Mobile Ltd. ( SEHK : 0941 and NYSE : CHL ) , the largest mobile carrier , and theGPS and social networking service provider , promoting GyPSii service before Beijing 2008 Olympic Games , according to thelatest news ."
2732,TN,"MILAN ( AFX ) - The Italian shipping group Grimaldi has increased its stake in the Finnish company Finnlines to 35.4 pct from the 30.5 pct held in June , according to news agency ANSA ."


In [14]:
aux_verb_df['Class'].value_counts()

Class
FP    1311
TN    1054
TP     319
FN      50
Name: count, dtype: int64

In [15]:
filt_nav = aux_verb_df['Class'] == "TP"
tp_nav_df = aux_verb_df[filt_nav]
tp_nav_df

,Class,Sentence
0,TP,"Master of Mayawas jointly developed by Nokia Siemens Networks and UFA - FremantleMedia , and will be actively advertised by Maxis in the end of May 2007 ."
1,TP,The acquisition will considerably increase Kemira 's sales and market position in the Russian metal industry coatings market .
2,TP,"The share sale , to foreign and Finnish investors , is expected to raise some euro300 million ( US$ 380 million ) ."
3,TP,"Later on , about 150 Finnish-made houses will be build in the area ."
4,TP,"In addition to fireplace exports , demand for lining stone has exceeded the level of the earlier part of the year and will continue to be clearly brisker for the remainder of the year ."
...,...,...
314,TP,"According shipping company Viking Line , the EU decision will have a significant financial impact ."
315,TP,"She will succeed Krister Kylas , who has decided to leave TeliaSonera ."
316,TP,TomTom has given assurances that it will continue to sell maps to competitors if it wins EU approval .
317,TP,"H+_kan Dahlstr+�m , head of mobility services at TeliaSonera , has forecast that mobile data volume on the TeliaSonera network in Sweden will rise eight-fold to 200,000 TB by 2014 ."


In [16]:
filt_nav = aux_verb_df['Class'] == "FN"
fn_nav_df = aux_verb_df[filt_nav]
fn_nav_df

,Class,Sentence
319,FN,The office space will rise above the remodeled Cannon Street underground station .
320,FN,"`` We are proud to welcome another distribution facility to north Mississippi and the region known as the logistics center of the United States , '' said Gray Swoope , executive director of MDA ."
321,FN,"Mr. Atul Chopra , Chief Operating Officer & President , Tecnotree elaborated '' I am eager to see the continued success of the former Tecnomen , bringing Lifetree product to MENA , under the Tecnotree umbrella ."
322,FN,"Reporting in accordance with the merged business operations will start as of January 1 , 2011 ."
323,FN,"Arto Ryymin , born 1964 , will replace Juhani Kaisanlahti who has worked as acting EVP , Healthcare & Welfare since August 2007 ."
324,FN,"UPM said the move will lower net profit by x20ac 385 million US$ 520 million in the second quarter , mainly due to impairment charges ."
325,FN,"The machine will have an annual production capacity of 200,000 tonnes of super-calendered magazine paper and other paper grades based on recovered fiber , Stora Enso said ."
326,FN,"By the end of 2006 , the number of joint branch offices will total 170 ."
327,FN,"Cencorp 's net sales in the first quarter is estimated to be EUR0.9-1 .2 m , as compared to EUR4 .5 m in the first quarter 2008 ."
328,FN,The transactions would increase earnings per share in the first quarter by some EUR0 .28 .
